src/test/fuzz/mini_miner.cpp

```diff
@@ -0,0 +1,183 @@
+namespace {
+
+const TestingSetup* g_setup;
+std::deque<COutPoint> g_available_coins;
+void initialize_miner()
+{
+    static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
+    g_setup = testing_setup.get();
+    for (uint32_t i = 0; i < uint32_t{100}; ++i) {
+        g_available_coins.push_back(COutPoint{uint256::ZERO, i});
+    }
+}
+
+// Test that the MiniMiner can run with various outpoints and feerates.
+FUZZ_TARGET_INIT(mini_miner, initialize_miner)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    CTxMemPool pool{CTxMemPool::Options{}};
+    std::vector<COutPoint> outpoints;
+    std::deque<COutPoint> available_coins = g_available_coins;
+    LOCK2(::cs_main, pool.cs);
+    // Cluster size cannot exceed 500
+    LIMITED_WHILE(!available_coins.empty(), 500)
+    {
+        CMutableTransaction mtx = CMutableTransaction();
+        const size_t num_inputs = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(1, available_coins.size());
+        const size_t num_outputs = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(1, 50);
+        for (size_t n{0}; n < num_inputs; ++n) {
+            auto prevout = available_coins.front();
+            mtx.vin.push_back(CTxIn(prevout, CScript()));
+            available_coins.pop_front();
+        }
+        for (uint32_t n{0}; n < num_outputs; ++n) {
+            mtx.vout.push_back(CTxOut(100, P2WSH_OP_TRUE));
+        }
+        CTransactionRef tx = MakeTransactionRef(mtx);
+        TestMemPoolEntryHelper entry;
+        const CAmount fee{ConsumeMoney(fuzzed_data_provider, /*max=*/MAX_MONEY/100000)};
+        assert(MoneyRange(fee));
+        pool.addUnchecked(entry.Fee(fee).FromTx(tx));
+
+        // All outputs are available to spend
+        for (uint32_t n{0}; n < num_outputs; ++n) {
+            if (fuzzed_data_provider.ConsumeBool()) {
+                available_coins.push_back(COutPoint{tx->GetHash(), n});
+            }
+        }
+
+        if (fuzzed_data_provider.ConsumeBool() && !tx->vout.empty()) {
+            // Add outpoint from this tx (may or not be spent by a later tx)
+            outpoints.push_back(COutPoint{tx->GetHash(),
+                                          (uint32_t)fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, tx->vout.size())});
+        } else {
+            // Add some random outpoint (will be interpreted as confirmed or not yet submitted
+            // to mempool).
+            auto outpoint = ConsumeDeserializable<COutPoint>(fuzzed_data_provider);
+            if (outpoint.has_value() && std::find(outpoints.begin(), outpoints.end(), *outpoint) == outpoints.end()) {
+                outpoints.push_back(*outpoint);
+            }
+        }
+
+    }
+
+    node::MiniMiner mini_miner{pool, outpoints};
+    assert(mini_miner.IsReadyToCalculate());
+    const CFeeRate target_feerate{CFeeRate{ConsumeMoney(fuzzed_data_provider, /*max=*/MAX_MONEY/1000)}};
+    if (fuzzed_data_provider.ConsumeBool()) {
```

Comment:
I was curious about a possible relation between `sum(CalculateBumpFees)` over all outpoints, and `CalculateTotalBumpFees` for the same outpoints / target feerate.
My naive thought was that taking the sum over `CalculateBumpFees` might overestimate the total bump fee by counting the bumping of some mempool txs twice if there is a shared ancestry, and that the quantities should be equal if there is no shared ancestry. As a result, `CalculateTotalBumpFees` should never be larger than the sum of `CalculateBumpFees`.
But when I tried to test this hypothesis by changing the fuzz test here to not calculate just one but both quantities in each run (using two mini miners) and then asserting that the above relation holds, it quickly fails - so where did I go wrong?

Reply:
Thanks, that’s unexpected. I would also expect that `CalculateTotalBumpFees(…)` always is equal or smaller to the sum of individual bump fees. I’ll check it out.

Reply:
I commented below for this. After fixing this, the fuzzer doesn't crash anymore (at least not immediately, haven't run it for long). 
I wonder if it might make sense to adjust the fuzz test permanently similar to my branch above and assert `sum(CalculateBumpFees) >= CalculateTotalBumpFees()`, if that is really the expectation - in general, I like it if fuzz tests contain asserts test actual non-trivial high-level expectations we have for the code, and don't just run over the code "blindly".

Reply:
Thanks, this improves the fuzzer. I’ve adopted your suggested change.